src/vm/virtualMachine.ts

```diff
@@ -351,4 +369,58 @@ export class VirtualMachine {
         break;
     }
   }
+
+  /** Relocate memory and trace */
+  relocate() {
+    let relocationTable = [1];
+    let memorySize = 1;
+
+    for (const segment of this.memory.values) {
+      memorySize += segment.length;
+      relocationTable.push(memorySize);
+    }
+
+    for (const [index, segment] of this.memory.values.entries()) {
+      for (const [offset, value] of segment.entries()) {
+        let relocatedValue: Felt;
+        const relocatedAddr = relocationTable[index] + offset;
+        if (isFelt(value)) {
+          relocatedValue = value;
+        } else if (isRelocatable(value)) {
+          relocatedValue = new Felt(
+            BigInt(relocationTable[value.segment] + value.offset)
+          );
+        } else {
+          throw new Error();
+        }
+
+        this.relocatedMemory.set(relocatedAddr, relocatedValue);
+      }
```

Comment:
then this is easier formulated as a simple `.flat` with a `.map` over each value to convert `Relocatable`

```
this.memory.values.flat().map((value) => isFelt(value) ? value : new Felt(
            BigInt(relocationTable[value.segment] + value.offset)
          )
```